### ISSUE TITLE
Mumudvb transcoding ffmpeg question

### ISSUE THREAD

**Issue by AlbertEinsteinGlitchPoint** [braice/MuMuDVB#216]
Hi guys

i have a strange issue..

i have tested mumudvb + ffmpeg 3.4.2 with ffserver configured all inside the same machine and it worked ok no problem..

but because CPU core was low for transcoding i have decided to setup another machine specific for transcoding with a 10x faster CPU with extra cores, but now i have a problem , i cannot get it to connect

mumudvb running channels in Multicast mode,  i have steup a NIC interface on transcoder server to connect no same switch multicast that its picking multicast channels from mumudvb , but nothing happens no errors i just star ffmpeg -i  udp://@XXXXXXXX  http://xxxxxxxxxx;port/channel/flv   but nothing happens

as if it was not receving the input stream from mumudvb.. this is very very strange..

does anyone have a specific external machine running transcoding externally?

if so how do i set it up?

**Comment by utelisysadmin** [comment]
Try pulling on the unicast. If it works troubleshoot your switch. Not all switches pass multicast by default.

**Comment by kslimani** [comment]
@AlbertEinsteinGlitchPoint on your transcoder, is your main network interface the multicast one ? or do you use a secondary network interface to connect to multicast ?

For example, if your configuration is the following :

main interface (unicast with some configured gateway) : `eth0`
secondary interface (multicast, no gateway) : `eth1`

In this scenario, you main needs to add the proper multicast route for the secondary interface.

For example, if your source is `udp://224.1.2.3:1234` you could manually ad a multicast route by typing :

```shell
route add -net 224.0.0.0 netmask 255.0.0.0 dev eth1
```
(adjust route ip/netmask to your needs)

**Comment by Saentist** [comment]
Use VLAN to transfer Multicast without to flood network

**Comment by kslimani** [comment]
I did not specify in my example, but i am of course assuming that network interfaces are connected on different vlan (eth0 unicast vlan and eth1 multicast vlan), and ideally using igmp in multicast vlan.

**Comment by AlbertEinsteinGlitchPoint** [comment]
Hi guys sorry for late reply

here is the scenario


PC 1  mumudvb running multicast Unicast and Multicast are working perfect  using 2 NIC interfaces

1 interface just for UNICAST
1 interface just for Multicast  ex: eth1

this eth1 outputs multicast traffic to network switch multicast..



PC2=  transcoder this pc receives the multicast and unicast stream from PC1 mumudvb

but so far i have not been able to connect both of them , i have ffserver running on PC2

and i have setup the following cmd

ffmpeg -i  udp://@239.106.1.0:1000  http://ip-of-pc2-transcoder:port/feed.ffm

it starts ffmpeg but does not find anything..

because i have 2 network interfaces how do i set ffmpeg to look into eth1 on PC2 ? this is the NIC where Multicast is reaching.. do i need to point this NIC interface on FFserver?



I am not worried about flood.. because this is still in testing environment

which means  Eth0 on both pcs is just for management, with multicast disabled on this eth0 inteface on both pcs..

and Eth1 is the one with multicast enabled on both pcs..

this eth1 interface both connect to a specific multicast switch ( this switch is not hooked up on any network, its just for usage on both pcs so share multicast between them 2.


Strange thing is i cannot get it to work, i will try to connect a Ethernet cable between both interfaces multicast without goint through the switch just for testing purposes..

**Comment by AlbertEinsteinGlitchPoint** [comment]
When i try to run Unicast on ffmpeg it gives me timeout connection here is output log

`root@encoder:~/ffserver# ffmpeg -i http://110.124.157.188:1016/bysid/1
ffmpeg version 3.4.2-1~16.04.york0.2 Copyright (c) 2000-2018 the FFmpeg developers
  built with gcc 5.4.0 (Ubuntu 5.4.0-6ubuntu1~16.04.9) 20160609
  configuration: --prefix=/usr --extra-version='1~16.04.york0.2' --toolchain=hardened --libdir=/usr/lib/x86_64-linux-gnu --incdir=/usr/include/x86_64-linux-gnu --enable-gpl --disable-stripping --enable-avresample --enable-avisynth --enable-gnutls --enable-ladspa --enable-libass --enable-libbluray --enable-libbs2b --enable-libcaca --enable-libcdio --enable-libflite --enable-libfontconfig --enable-libfreetype --enable-libfribidi --enable-libgme --enable-libgsm --enable-libmp3lame --enable-libmysofa --enable-libopenjpeg --enable-libopenmpt --enable-libopus --enable-libpulse --enable-librubberband --enable-librsvg --enable-libshine --enable-libsnappy --enable-libsoxr --enable-libspeex --enable-libssh --enable-libtheora --enable-libtwolame --enable-libvorbis --enable-libvpx --enable-libwavpack --enable-libwebp --enable-libx265 --enable-libxml2 --enable-libxvid --enable-libzmq --enable-libzvbi --enable-omx --enable-openal --enable-opengl --enable-sdl2 --enable-libdc1394 --enable-libdrm --enable-libiec61883 --enable-chromaprint --enable-frei0r --enable-libopencv --enable-libx264 --enable-shared
  libavutil      55. 78.100 / 55. 78.100
  libavcodec     57.107.100 / 57.107.100
  libavformat    57. 83.100 / 57. 83.100
  libavdevice    57. 10.100 / 57. 10.100
  libavfilter     6.107.100 /  6.107.100
  libavresample   3.  7.  0 /  3.  7.  0
  libswscale      4.  8.100 /  4.  8.100
  libswresample   2.  9.100 /  2.  9.100
  libpostproc    54.  7.100 / 54.  7.100
http://110.124.157.188:1016/bysid/1: Connection timed out
`

**Comment by kslimani** [comment]
@AlbertEinsteinGlitchPoint please read my previous answer, i think you need to add multicast route to secondary network interface on PC2.

EDIT:

try typing this on PC2 (as root user and assuming secondary network interface for multicast is eth1):

```
route add -net 239.0.0.0 netmask 255.0.0.0 dev eth1
```
EDIT2: note also this adds the multicast route manually and will not persists on server reboot

**Comment by AlbertEinsteinGlitchPoint** [comment]
Hi utelisysadmin

it did not work neither on unicast stream...

this is very strange.. on multicast it does not timeout at all  ffmpeg keeps running but does not transcode anything..

using bmon tool i can see my multicast eth1 interface on PC2 transcoder receiving the multicast stream at least its passing around 20Mbps traffic from the hdtv full hd 1920x1080p channel being transmitted on multicast.

**Comment by AlbertEinsteinGlitchPoint** [comment]
This what shows when starting ffmpeg on multicast on PC2

`root@encoder:~/ffserver# ffmpeg -i udp://@239.106.1.0:1000
ffmpeg version 3.4.2-1~16.04.york0.2 Copyright (c) 2000-2018 the FFmpeg developers
  built with gcc 5.4.0 (Ubuntu 5.4.0-6ubuntu1~16.04.9) 20160609
  configuration: --prefix=/usr --extra-version='1~16.04.york0.2' --toolchain=hardened --libdir=/usr/lib/x86_64-linux-gnu --incdir=/usr/include/x86_64-linux-gnu --enable-gpl --disable-stripping --enable-avresample --enable-avisynth --enable-gnutls --enable-ladspa --enable-libass --enable-libbluray --enable-libbs2b --enable-libcaca --enable-libcdio --enable-libflite --enable-libfontconfig --enable-libfreetype --enable-libfribidi --enable-libgme --enable-libgsm --enable-libmp3lame --enable-libmysofa --enable-libopenjpeg --enable-libopenmpt --enable-libopus --enable-libpulse --enable-librubberband --enable-librsvg --enable-libshine --enable-libsnappy --enable-libsoxr --enable-libspeex --enable-libssh --enable-libtheora --enable-libtwolame --enable-libvorbis --enable-libvpx --enable-libwavpack --enable-libwebp --enable-libx265 --enable-libxml2 --enable-libxvid --enable-libzmq --enable-libzvbi --enable-omx --enable-openal --enable-opengl --enable-sdl2 --enable-libdc1394 --enable-libdrm --enable-libiec61883 --enable-chromaprint --enable-frei0r --enable-libopencv --enable-libx264 --enable-shared
  libavutil      55. 78.100 / 55. 78.100
  libavcodec     57.107.100 / 57.107.100
  libavformat    57. 83.100 / 57. 83.100
  libavdevice    57. 10.100 / 57. 10.100
  libavfilter     6.107.100 /  6.107.100
  libavresample   3.  7.  0 /  3.  7.  0
  libswscale      4.  8.100 /  4.  8.100
  libswresample   2.  9.100 /  2.  9.100
  libpostproc    54.  7.100 / 54.  7.100
`

**Comment by kslimani** [comment]
To check if your PC2 can pull the stream from PC1 using multicast, use ffprobe tool :

```
ffprobe udp://239.106.1.0:1000
```
Note: do not use the "@" notation it is not needed with FFmpeg (only with VLC i think)

**Comment by AlbertEinsteinGlitchPoint** [comment]
Hi kslimani

PC2 eth1 has the same route add , i already done it before

and ffprobe gets stuck on 

> root@encoder:~# ffprobe udp://239.106.0.0:1000
ffprobe version 3.4.2-1~16.04.york0.2 Copyright (c) 2007-2018 the FFmpeg develop                                                                                                                     ers
  built with gcc 5.4.0 (Ubuntu 5.4.0-6ubuntu1~16.04.9) 20160609
  configuration: --prefix=/usr --extra-version='1~16.04.york0.2' --toolchain=har                                                                                                                     dened --libdir=/usr/lib/x86_64-linux-gnu --incdir=/usr/include/x86_64-linux-gnu                                                                                                                      --enable-gpl --disable-stripping --enable-avresample --enable-avisynth --enable-                                                                                                                     gnutls --enable-ladspa --enable-libass --enable-libbluray --enable-libbs2b --ena                                                                                                                     ble-libcaca --enable-libcdio --enable-libflite --enable-libfontconfig --enable-l                                                                                                                     ibfreetype --enable-libfribidi --enable-libgme --enable-libgsm --enable-libmp3la                                                                                                                     me --enable-libmysofa --enable-libopenjpeg --enable-libopenmpt --enable-libopus                                                                                                                      --enable-libpulse --enable-librubberband --enable-librsvg --enable-libshine --en                                                                                                                     able-libsnappy --enable-libsoxr --enable-libspeex --enable-libssh --enable-libth                                                                                                                     eora --enable-libtwolame --enable-libvorbis --enable-libvpx --enable-libwavpack                                                                                                                      --enable-libwebp --enable-libx265 --enable-libxml2 --enable-libxvid --enable-lib                                                                                                                     zmq --enable-libzvbi --enable-omx --enable-openal --enable-opengl --enable-sdl2                                                                                                                      --enable-libdc1394 --enable-libdrm --enable-libiec61883 --enable-chromaprint --e                                                                                                                     nable-frei0r --enable-libopencv --enable-libx264 --enable-shared
  libavutil      55. 78.100 / 55. 78.100
  libavcodec     57.107.100 / 57.107.100
  libavformat    57. 83.100 / 57. 83.100
  libavdevice    57. 10.100 / 57. 10.100
  libavfilter     6.107.100 /  6.107.100
  libavresample   3.  7.  0 /  3.  7.  0
  libswscale      4.  8.100 /  4.  8.100
  libswresample   2.  9.100 /  2.  9.100
  libpostproc    54.  7.100 / 54.  7.100

**Comment by AlbertEinsteinGlitchPoint** [comment]
ok guys further testing i found something more bizzare..

i have multicast traffic comming trough on the enconder PC2 but it seems its comming via the IPv4 valid interface , weird..



`root@encoder:~# tcpdump -i enp4s0
tcpdump: verbose output suppressed, use -v or -vv for full protocol decode
listening on enp4s0, link-type EN10MB (Ethernet), capture size 262144 bytes
20:48:13.261148 IP 170.144.137.108.47838 > 239.106.32.0.1000: UDP, length 1316
20:48:13.261156 IP 170.144.137.108.47838 > 239.106.32.0.1000: UDP, length 1316
20:48:13.261161 IP 170.144.137.108.47838 > 239.106.32.0.1000: UDP, length 1316
20:48:13.261351 IP 170.144.137.108.47838 > 239.106.32.0.1000: UDP, length 1316
20:48:13.261599 IP 170.144.137.108.47838 > 239.106.32.0.1000: UDP, length 1316
20:48:13.261607 IP 170.144.137.108.47838 > 239.106.32.0.1000: UDP, length 1316
20:48:13.261612 IP 170.144.137.108.47838 > 239.106.32.0.1000: UDP, length 1316
20:48:13.457230 IP 170.144.137.108.47838 > 239.106.224.0.1000: UDP, length 1316



BTW 170.144.137.108:47838   this is the IPV4 on the Eth0 of PC1 middlware..

strange it seems its beeing sent from it..

but i have disabled multicast on that interface using ifconfig eth0 -multicast


`

**Comment by AlbertEinsteinGlitchPoint** [comment]
weird i have multicast stream arriving at PC2 now,, but ffprobe and ffmpeg -i cannot detect this stream how weird is that lol

**Comment by Saentist** [comment]
Is there support for multicast in firewall and kernel?

**Comment by kslimani** [comment]
@AlbertEinsteinGlitchPoint i think the issue comes from your PC1. If the main network interface on this computer is the unicast one and the secondary the multicast one, then you need also to add a multicast route to secondary interface.

On the PC1 type also the following command :

```
route add -net 239.0.0.0 netmask 255.0.0.0 dev eth1
```
Note: replace "eth1" by the secondary network interface device name (and adjust ip/netmask to your needs).

**Comment by AlbertEinsteinGlitchPoint** [comment]
Hi Kslimani i already done that added the route -net to eth1 which is the multicast interface

netstat  shows the route added on pc1 and pc2  on the specific NIC multicast for both machines

saentist how do i check the route on the kernel?  firewall is disabled in both machines they are running inside same local network

**Comment by AlbertEinsteinGlitchPoint** [comment]
ok to clear things up

this is my netstat -g output on PC1 server middleware

enp3s4f0 - is my ethernet which is running multicast from mumudvb

 netstat -g
IPv6/IPv4 Group Memberships
Interface       CntRef Grupo
--------------- ------ ---------------------
lo              1      all-systems.mcast.net
enp0s31f6       1      all-systems.mcast.net
enp3s4f0        1      239.106.32.0
enp3s4f0        1      all-systems.mcast.net
lo              1      ip6-allnodes
lo              1      ff01::1
enp2s4f0        1      ip6-allnodes
enp2s4f0        1      ff01::1
enp0s31f6       1      ff02::1:ff71:1e9e
enp0s31f6       1      ip6-allnodes
enp0s31f6       1      ff01::1
enp2s4f1        1      ip6-allnodes
enp2s4f1        1      ff01::1
enp3s4f0        1      ff02::1:ff08:224a
enp3s4f0        1      ip6-allnodes
enp3s4f0        1      ff01::1
enp3s4f1        1      ip6-allnodes
enp3s4f1        1      ff01::1
enp3s5f0        1      ip6-allnodes
enp3s5f0        1      ff01::1
enp3s5f1        1      ip6-allnodes
enp3s5f1        1      ff01::1



and this is the output on PC2 transcoder  

the multicast ethernet card on this pc2 transcoder is enp4s0

root@encoder:~# netstat -g
IPv6/IPv4 Group Memberships
Interface       RefCnt Group
--------------- ------ ---------------------
lo              1      all-systems.mcast.net
enp2s0          1      all-systems.mcast.net
enp4s0          1      all-systems.mcast.net
docker0         1      all-systems.mcast.net
lo              1      ip6-allnodes
lo              1      ff01::1
enp2s0          1      ff02::1:ff01:f397
enp2s0          1      ip6-allnodes
enp2s0          1      ff01::1
enp4s0          1      ff02::fb
enp4s0          1      ff02::1:ff03:c379
enp4s0          1      ip6-allnodes
enp4s0          1      ff01::1
docker0         1      ip6-allnodes
docker0         1      ff01::1
root@encoder:~#

**Comment by AlbertEinsteinGlitchPoint** [comment]
Hi guys ok after reboot the PC2 , all the sudden it started working the INPUT multicast arriving

but now o still another weird issue, i fireup ffserver, then i run ffmpeg -i  it starts.. but just before starting transcoding it gives me another TCP connection refused on the port choosen..

[mpegts @ 0x55bdec763f40] PES packet size mismatch
[mp3 @ 0x55bdec78b5c0] Header missing
    Last message repeated 1 times
[mpegts @ 0x55bdec763f40] PES packet size mismatch
[mp3 @ 0x55bdec78b5c0] Header missing
[mpegts @ 0x55bdec763f40] decoding for stream 2 failed
[mpegts @ 0x55bdec763f40] Could not find codec parameters for stream 2 (Audio: mp3 ([17][0][0][0] / 0x0011), 0 channels, s16p): unspecified frame size
Consider increasing the value for the 'analyzeduration' and 'probesize' options
[mpegts @ 0x55bdec763f40] Could not find codec parameters for stream 4 (Unknown: none ([11][0][0][0] / 0x000B)): unknown codec
Consider increasing the value for the 'analyzeduration' and 'probesize' options
[mpegts @ 0x55bdec763f40] Could not find codec parameters for stream 5 (Unknown: none ([5][0][0][0] / 0x0005)): unknown codec
Consider increasing the value for the 'analyzeduration' and 'probesize' options
Input #0, mpegts, from 'udp://@239.106.32.0:1000':
  Duration: N/A, start: 61677.213167, bitrate: N/A
  Program 13600
    Metadata:
      service_name    : ?TVM HDTV
      service_provider:
    Stream #0:0[0x111]: Video: h264 (High) ([27][0][0][0] / 0x001B), yuv420p(tv, bt709, top first), 1920x1080 [SAR 1:1 DAR 16:9], 29.97 fps, 59.94 tbr, 90k tbn, 59.94 tbc
    Stream #0:1[0x112]: Audio: aac_latm (HE-AAC) ([17][0][0][0] / 0x0011), 48000 Hz, stereo, fltp
    Stream #0:2[0x113]: Audio: mp3 ([17][0][0][0] / 0x0011), 0 channels, s16p
    Stream #0:3[0x116]: Data: bin_data ([6][0][0][0] / 0x0006)
    Stream #0:4[0x384]: Unknown: none ([11][0][0][0] / 0x000B)
    Stream #0:5[0x1f4]: Unknown: none ([5][0][0][0] / 0x0005)
[tcp @ 0x55bdec794d00] Connection to tcp://170.144.223.244:8892 failed: Connection refused
http://170.144.223.244:8892/tvg1.ffm: Connection refused



I am running out of ideas now.

**Comment by AlbertEinsteinGlitchPoint** [comment]
after long tries, and endless reboots on the PC2 transcoder

finally i have managed to start transcoding the input multicast stream, but now channels are freezing like crazy hell every second or so..  which does not happen on the same transcoding channel using ffmpeg on same direct machine as mumudvb is running.. now i am out of ideas the only thing i could se on ffmpeg log on pc2 transcoder is alot of error PES size mismatch


[h264 @ 0x559d0ed4b140] cabac decode of qscale diff failed at 58 22
[h264 @ 0x559d0ed4b140] error while decoding MB 58 22, bytestream 5606
[h264 @ 0x559d0ed4b140] concealing 5511 DC, 5511 AC, 5511 MV errors in P frame
[mpegts @ 0x559d0ebe3f40] PES packet size mismatch
[h264 @ 0x559d0ec9f260] concealing 3352 DC, 3352 AC, 3352 MV errors in B frame
[h264 @ 0x559d0ec791c0] Reference 2 >= 2
[h264 @ 0x559d0ec791c0] error while decoding MB 49 24, bytestream 5755
[h264 @ 0x559d0ec791c0] concealing 5280 DC, 5280 AC, 5280 MV errors in P frame
[aac_latm @ 0x559d0ecdb5e0] Number of bands (55) exceeds limit (42).
Error while decoding stream #0:1: Invalid data found when processing input
    Last message repeated 1 times
[mpegts @ 0x559d0ebe3f40] PES packet size mismatch
[h264 @ 0x559d0ec9f260] Reference 2 >= 2
[h264 @ 0x559d0ec9f260] error while decoding MB 72 30, bytestream 17244
[h264 @ 0x559d0ec9f260] concealing 4537 DC, 4537 AC, 4537 MV errors in P frame
[mpegts @ 0x559d0ebe3f40] PES packet size mismatch
[aac_latm @ 0x559d0ecdb5e0] Number of bands (53) exceeds limit (42).
Error while decoding stream #0:1: Invalid data found when processing input
[h264 @ 0x559d0eee7660] Reference 4 >= 2
[h264 @ 0x559d0eee7660] error while decoding MB 69 62, bytestream 20013
[h264 @ 0x559d0eee7660] concealing 700 DC, 700 AC, 700 MV errors in B frame
[mpegts @ 0x559d0ebe3f40] PES packet size mismatch
[h264 @ 0x559d0ece97a0] concealing 859 DC, 859 AC, 859 MV errors in B framebits/s dup=1303 drop=2 speed=1.06x
[mpegts @ 0x559d0ebe3f40] PES packet size mismatch
Error while decoding stream #0:1: Invalid data found when processing input
[mpegts @ 0x559d0ebe3f40] PES packet size mismatch
[h264 @ 0x559d0ed4b140] Reference 4 >= 2
[h264 @ 0x559d0ed4b140] error while decoding MB 116 16, bytestream 9633
[h264 @ 0x559d0ed4b140] concealing 6173 DC, 6173 AC, 6173 MV errors in B frame
Error while decoding stream #0:1: Invalid data found when processing input
[h264 @ 0x559d0ec9f260] Reference 2 >= 2
[h264 @ 0x559d0ec9f260] error while decoding MB 112 8, bytestream 92768
[h264 @ 0x559d0ec9f260] concealing 7137 DC, 7137 AC, 7137 MV errors in B frame

**Comment by utelisysadmin** [comment]
Most likely network related. Put an input buffer on ffmpeg or try to solve the network issue. Also you should most likely be receiving udp datagrams of size 1316, You can also check that with tcpdump.

**Comment by AlbertEinsteinGlitchPoint** [comment]
hi utelisysadmin

ok yes on tcpdump i see packets all in 1316 byte size, but channels are freezing alot..

i also tried connecting both machines together with patch cable and also on tcpdump i see same 1316 packets multicast arriving but ffmpeg cannot filter stream now..

its very very weird.

**Comment by AlbertEinsteinGlitchPoint** [comment]
btw how to i put a input buffer on ffmpeg? what kind of config shall i use?


i am getting crazy with this stuff... if i fire up   ffmpeg with ffserver on same pc server from mumudvb transcoding works flawless.. no problems

when i fire up second pc on the network, to run another ffmpeg instace to transcode more channels 

i get 50/50 rate some times i cannot get any multicast input route..

when i add route to pc transcoding machine i star receiving multicast packets all with 1316 size..

but when i fireup ffprobe or ffmpeg no input data is reached to transcoder..

but on tcpdump i see all the multicast packets arriving the NIC multicast interface.

and when ffmpeg and ffprobe detects this multicast stream, on transcoding the channel it glitches all the time, picture freezing  all the time, its unwatcheable.

**Comment by utelisysadmin** [comment]
I guess you need to add something along the lines of:

-i "http://ip_or_hostname:port_number?fifo_size=1000000&overrun_nonfatal=1"

You could also increase the logging level and check if there are any other useful messages.
"-loglevel debug" right after the "ffmpeg" command.
